File: 0011-container-with-most-water/0011-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
    // Initialize two pointers, one at the start (i) and one at the end (j)
    let i = 0;
    let j = height.length - 1;
    
    // Variable to store the maximum area
    let ans = 0;
    
    // Loop until the two pointers meet
    while (i < j) {
        // Calculate the current area using the shorter height and distance between the two pointers
        const t = Math.min(height[i], height[j]) * (j - i);
        
        // Update the maximum area if the current area is larger
        ans = Math.max(ans, t);
        
        // Move the pointer with the shorter height inward
        if (height[i] < height[j]) {
            ++i; // Move left pointer right
        } else {
            --j; // Move right pointer left
        }
    }
    
    // Return the maximum area found
    return ans;
};

File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        // Initialize two pointers, i at the start and j at the end of the array
        int i = 0, j = height.length - 1;
        // Variable to store the maximum area found
        int ans = 0;
        
        // Loop until the two pointers meet
        while (i < j) {
            // Calculate the area by using the shorter height and the distance between the two pointers
            int t = Math.min(height[i], height[j]) * (j - i);
            // Update the maximum area found so far
            ans = Math.max(ans, t);
            
            // Move the pointer that has the shorter height to try to find a larger area
            if (height[i] < height[j]) {
                ++i; // Move the left pointer to the right
            } else {
                --j; // Move the right pointer to the left
            }
        }
        
        // Return the maximum area
        return ans;
    }
}
